### ISSUE TITLE
--experimental_output_paths=strip is not effective when actions are scheduled in parallel

### ISSUE THREAD

**Issue by DavidANeil** [bazelbuild/bazel#21043]
### Description of the feature request:

`--experimental_output_paths=strip` can allow for a `exec` and `target` spawn to have the same input digest, and therefore be cache hits.  Unfortunately, because these actions frequently have identical or exactly duplicated input dependency graphs, it isn't uncommon that the two actions are scheduled at the same time, meaning the work is still duplicated.

I'm not totally certain on the solution, but my first bad idea is that the spawn strategy should, if this flag is enabled, check to see if any sibling spawns match the input digest of spawn that is about to start executing, and if so somehow "join" the spawn results.

### Which category does this issue belong to?

Local Execution

### What underlying problem are you trying to solve with this feature?

Our host and target platforms are identical, so all duplicated `[for tool]` actions are pure overhead.  This flag seems promising at reducing that overhead, but we still frequently see duplicated actions since the "exec" and "target" dependency chains tend to stay in sync, since one branch falling behind due to chance of the scheduler means it gets cache hits until it catches back up and starts duplicating execution again.

**Comment by fmeum** [comment]
CC @coeuvre @tjgq @gregestren If I understand the logic around `RemoteCacheClient` correctly, Bazel will not deduplicate inflight requests to the remote cache/executor - which makes sense without path mapping, as an action previously couldn't generate cache hits for other actions in the same build. Is this something you think could be changed easily, e.g. by tracking in-flight requests and returning a `Future` waiting for the result of the first one?

**Comment by fmeum** [comment]
To be more concrete, this is the test case that we would want to pass:
```diff
diff --git a/src/test/shell/bazel/path_mapping_test.sh b/src/test/shell/bazel/path_mapping_test.sh
index 3014ba8878..e53ee3876d 100755
--- a/src/test/shell/bazel/path_mapping_test.sh
+++ b/src/test/shell/bazel/path_mapping_test.sh
@@ -286,4 +286,67 @@ EOF
   expect_log '2 remote[^ ]'
 }
 
+function test_path_stripping_cache_hit_for_parallel_action() {
+  mkdir rules
+  cat > rules/defs.bzl <<'EOF'
+def _platforms_split_transition_impl(settings, attr):
+    return [
+        {"//command_line_option:platforms": "//rules:foo"},
+        {"//command_line_option:platforms": "//rules:bar"},
+    ]
+
+_platforms_split_transition = transition(
+    implementation = _platforms_split_transition_impl,
+    inputs = [],
+    outputs = ["//command_line_option:platforms"],
+)
+
+def _cat_impl(ctx):
+    out = ctx.actions.declare_file(ctx.label.name)
+    inputs = [file for src in ctx.attr.src for file in src[DefaultInfo].files.to_list()]
+    ctx.actions.run_shell(
+        inputs = inputs,
+        outputs = [out],
+        command = "cat {} > {}".format(" ".join([f.path for f in inputs]), out.path),
+    )
+    return [
+        DefaultInfo(files = depset([out])),
+    ]
+
+cat = rule(
+    _cat_impl,
+    attrs = {
+        "src": attr.label(cfg = _platforms_split_transition),
+    },
+)
+EOF
+  cat > rules/BUILD << 'EOF'
+platform(name = "foo", parents = ["@local_config_platform//:host"])
+platform(name = "bar", parents = ["@local_config_platform//:host"])
+EOF
+
+  mkdir -p pkg
+  cat > pkg/BUILD <<'EOF'
+load("//rules:defs.bzl", "cat")
+
+genrule(
+    name = "gen_src",
+    outs = ["gen_src.txt"],
+    cmd = "echo 'Hello, World!' > $@",
+)
+
+cat(
+    name = "cat",
+    src = ":gen_src",
+)
+EOF
+
+  bazel build \
+    --experimental_output_paths=strip \
+    --remote_executor=grpc://localhost:${worker_port} \
+    //pkg:cat &> $TEST_log || fail "build failed unexpectedly"
+  expect_log '2 remote[^ ]'
+  expect_log '1 remote cache hit'
+}
+
 run_suite "path mapping tests"
```

**Comment by tjgq** [comment]
> @coeuvre @tjgq If I understand the logic around `RemoteCacheClient` correctly, Bazel will not deduplicate inflight requests to the remote cache/executor - which makes sense without path mapping, as an action previously couldn't generate cache hits for other actions in the same build. Is this something you think could be changed easily, e.g. by tracking in-flight requests and returning a `Future` waiting for the result of the first one?

I think this makes sense, yes.

**Comment by tjgq** [comment]
Thinking some more about it: is this really worth doing if we assume that the remote executor is able to deduplicate identical concurrent requests on its own? The spec doesn't mandate it, but it does [hint](https://cs.opensource.google/bazel/bazel/+/master:third_party/remoteapis/build/bazel/remote/execution/v2/remote_execution.proto;l=1355;drc=e382abf33f1aedc18f19e9bf2bb2e407035f2c92) at the possibility, and I'd expect a well-optimized implementation to do so. (It's at least true of Google's internal equivalent of remote execution.)

On the Bazel side, we can't really save any pre-execution work (we need to calculate the action digest anyway to detect duplicates) nor post-execution work (we don't have the ability to deduplicate downloads of the same blob into many output locations, and even if we did, it would only matter for actions "built with the bytes"). So I think the gains would be solely in reduced (non-blob) network traffic, which I'm not sure is worth the extra complexity.

**Comment by fmeum** [comment]
That's a good point! Since deduplication within a single build is really just a special case of general deduplication across concurrent builds for a remote executor, it's better positioned to do this than Bazel.

However, this still leaves the case of local execution with a remote (or disk) cache. In this situation, I don't see how it would be possible for the cache to transparently deduplicate requests, given that it doesn't have control over whether and when the results of an action that incurred a cache miss will be uploaded. 

@tjgq Do you maybe see a way to push this kind of deduplication into the cache? If not, would you consider that to be a sensible addition?

@DavidANeil I just noticed the "Local execution" category you chose for the issue. Are you using a remote cache together with local execution?

**Comment by tjgq** [comment]
For local execution, I think the rough outline of "keep a map of futures for in-flight executions, resolve the future on action upload, block on the future when downloading an identical action" might be doable. But it's the sort of thing that I'd have to write to convince myself (for example, it's a little unclear to me what the failure path would look like).

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.3.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.3.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.3.0rc1`. Thanks!
